### PR TITLE
Introduce Y022/Y023/Y024/Y025: Imports linting error codes

### PR DESCRIPTION
**Pull request by AlexWaygood**
Helps work towards #80 and #32. I've tried to keep these error codes completely cross-compatible between Python 2 and Python 3 -- I think it would be best to leave Python-3-only changes (e.g. disallowing `typing.ContextManager` and `typing_extensions.OrderedDict`) can be left for another error code.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -81,6 +81,14 @@ currently emitted:
   an instance of ``cls``, and ``__new__`` methods.
 * Y020: Quoted annotations should never be used in stubs.
 * Y021: Docstrings should not be included in stubs.
+* Y022: Imports linting: use typing-module aliases to stdlib objects as little
+  as possible (e.g. use ``builtins.list`` instead of ``typing.List``). However,
+  use ``typing.NamedTuple`` instead of ``collections.namedtuple``, as it allows
+  for more precise type inference. For typing-only objects such as ``ClassVar``
+  or ``NoReturn``, import objects from ``typing`` instead of
+  ``typing_extensions`` wherever possible.
```
Comment (JelleZijlstra) [inline README.rst:89]: I feel like this should be split in three:
- one code for PEP 585 stuff (builtins instead of typing)
- one for NamedTuple
- one for using typing instead of typing-extensions where applicable

Comment (AlexWaygood) [reply README.rst:89]: It's annoying because from the perspective of writing the code, it's the same check. But I see what you mean, I'll revise it 🙂

**Comment by AlexWaygood** [comment]
Okay, should be ready for review now. I'll hold off running a codemodding script on the Python-2 parts of typeshed until people have had a chance to review :)

File: pyi.py [hunk 2]
```diff
@@ -42,6 +42,58 @@ class TypeVarInfo(NamedTuple):
...
 
+# The collections import blacklist is for typing & typing_extensions
+#
+# OrderedDict is omitted:
+# -- In Python 3, we'd rather import it from collections, not typing or typing_extensions
+# -- But in Python 2, it cannot be imported from collections or typing, only from typing_extensions
+#
+# ChainMap does not exist in typing or typing_extensions in Python 2,
+# so we can disallow importing it from anywhere except collections
+_BAD_COLLECTIONS_ALIASES = {
```
Comment (JelleZijlstra) [inline pyi.py:53]: These names are kind of unclear, could you name them similarly to _TYPING_NOT_TYPING_EXTENSIONS with both the module they should be imported from and the module they should *not* be imported from?

Comment (AlexWaygood) [reply pyi.py:53]: sure

File: pyi.py [hunk 3]
```diff
@@ -187,6 +239,76 @@ def in_class(self) -> bool:
...
+            bad_cls_alias=f"{module_name}.{object_name}",
+        )
+        self.error(node, error_message)
+
+    def visit_Attribute(self, node: ast.Attribute) -> None:
+        self.generic_visit(node)
+        thing = node.value
+        if not isinstance(thing, ast.Name):
+            return
+        thingname, attribute = thing.id, node.attr
```
Comment (JelleZijlstra) [inline pyi.py:279]:
```suggestion
        module_name, attribute = thing.id, node.attr
```
More precise variable name

File: pyi.py [hunk 4]
```diff
@@ -187,6 +239,76 @@ def in_class(self) -> bool:
...
+                return
+            error_code, blacklist = Y022, _BAD_BUILTINS_ALIASES
+        elif object_name in _BAD_COLLECTIONS_ABC_ALIASES:
+            error_code, blacklist = Y023, _BAD_COLLECTIONS_ABC_ALIASES
+        elif object_name in _TYPING_NOT_TYPING_EXTENSIONS:
+            error_code, blacklist = Y023, _TYPING_NOT_TYPING_EXTENSIONS
+        elif object_name == "ContextManager":
+            error_code = Y023
+            blacklist = {
+                "ContextManager": '"contextlib.AbstractContextManager" or "typing.ContextManager"'
```
Comment (JelleZijlstra) [inline pyi.py:263]: Shouldn't we pick one to recommend?

File: pyi.py (hunk 3 above)
Comment (AlexWaygood) [reply pyi.py:279]: I went for `thingname` because we *don't* know that it's a module here -- this method visits all `ast.Attribute` nodes. It's only later down the method that we narrow it down as to whether it's an `ast.Attribute` we're interested in or not.

File: pyi.py [hunk 5]
```diff
@@ -187,6 +239,76 @@ def in_class(self) -> bool:
         """Determine whether we are inside a `class` statement"""
         return bool(self._class_nesting)
 
+    def _check_typing_object(
+        self, node: ast.Attribute | ast.ImportFrom, module_name: str, object_name: str
+    ) -> None:
+        if object_name in _BAD_COLLECTIONS_ALIASES:
```
Comment (JelleZijlstra) [inline pyi.py:245]: This logic is a bit hard to follow because we don't always check the module name. You have to look at the other methods to see that this gets called only for typing/typing_extensions.

What if instead we always called _check_typing_object for each attribute and import, and then here had logic like
```
if module_name == "typing":
    if object_name not in _BAD_BUILTINS_ALIASES:
        return
    replacement = ...
elif module_name == "typing_extensions":
     # different logic for deciding if the name is bad
self.error(...)
```
That makes it easier to see at a glance which names will produce errors for each module.

File: pyi.py (hunk 4 above)
Comment (AlexWaygood) [reply pyi.py:263]: `contextlib.AbstractContextManager` will fail on Python 2, typeshed still supports Python 2

File: pyi.py (hunk 2 above)
Comment (AlexWaygood) [reply pyi.py]: Done.

File: pyi.py (hunk 4 above)
Comment (JelleZijlstra) [reply pyi.py:263]: I see, then maybe we should say something like `'"contextlib.AbstractContextManager" (or "typing.ContextManager" in Python 2-compatible code)'`. When a user sees this error message, I want them to have enough information to know how to fix it, and if we give them two names, how are they going to know which one to use?

File: pyi.py (hunk 5 above)
Comment (AlexWaygood) [reply pyi.py]: Okay!

File: pyi.py (hunk 2 above)
Comment (JelleZijlstra) [reply pyi.py]: Thanks!

File: pyi.py (hunk 4 above)
Comment (AlexWaygood) [reply pyi.py:263]: That makes sense... What do you think about the `collections.abc` case, though? `use "collections.abc.Awaitable", unless you stumble across a random mypy/pytype bug, in which case use `typing.Awaitable`?

Comment (JelleZijlstra) [reply pyi.py:263]: I guess we should fix those bugs so this isn't an issue any more :) I don't recall the details of those bugs, are they common? Maybe we should just recommend typing until it's always safe to use collections.abc.

Comment (AlexWaygood) [reply pyi.py:263]: > I guess we should fix those bugs so this isn't an issue any more :)

Indeed 😀

>  I don't recall the details of those bugs, are they common?

- https://github.com/python/mypy/issues/11860
- https://github.com/google/pytype/issues/1096

Comment (AlexWaygood) [reply pyi.py:263]: TBF, the mypy issue looks typeshed-specific, and the pytype may be as well. But we certainly can't yet run a codemodding script using `collections.abc` everywhere in typeshed (if we could, I would have already done so ;)

File: README.rst [hunk 6]
```diff
@@ -81,6 +81,16 @@ currently emitted:
...
 * Y021: Docstrings should not be included in stubs.
+* Y022: Imports linting: use typing-module aliases to stdlib objects as little
+  as possible (e.g. ``builtins.list`` over ``typing.List``,
+  ``collections.Counter`` over ``typing.Counter``, etc.).
+* Y023: Where there is no detriment to backwards compatibility, import objects
+  such as ``ClassVar`` and ``NoReturn`` from ``typing`` rather than
+  ``typing_extensions``.
+* Y024: Use ``typing.NamedTuple`` instead of ``collections.namedtuple``, as it allows
+  for more precise type inference.
+* Y023: Always alias ``collections.abc.Set`` when importing it, so as to avoid
```
Comment (JelleZijlstra) [inline README.rst:92]:
```suggestion
* Y025: Always alias ``collections.abc.Set`` when importing it, so as to avoid
```

Comment (AlexWaygood) [reply README.rst:92]: Good catch!

File: pyi.py [hunk 7]
```diff
@@ -753,6 +881,14 @@ def should_warn(self, code):
...
 Y020 = "Y020 Quoted annotations should never be used in stubs"
 Y021 = "Y021 Docstrings should not be included in stubs"
+Y022 = 'Y022 Use {good_cls_name} instead of "{bad_cls_alias}"'
+Y023 = 'Y023 Use {good_cls_name} instead of "{bad_cls_alias}"'
+Y024 = 'Y024 Use "typing.NamedTuple" instead of "collections.namedtuple"'
+Y025 = (
+    'Y025 Use "from collections.abc import Set as AbstractSet" '
+    'to avoid confusion with "builtins.set"'
+)
+Y092 = "Y092 Top-level attribute must not have a default value"
```
Comment (JelleZijlstra) [inline pyi.py:891]:
```suggestion
```
Bad merge

File: pyi.py [hunk 8]
```diff
@@ -187,6 +239,81 @@ def in_class(self) -> bool:
         """Determine whether we are inside a `class` statement"""
         return bool(self._class_nesting)
 
+    def _check_import_or_attribute(
+        self, node: ast.Attribute | ast.ImportFrom, module_name: str, object_name: str
+    ) -> None:
```
Comment (JelleZijlstra) [inline pyi.py:244]: Thanks, this is much easier to read now

**Review by JelleZijlstra — APPROVED** [review]

**Comment by AlexWaygood** [comment]
I will get codemodding tomorrow!

File: pyi.py [hunk 9]
```diff
@@ -42,6 +42,58 @@ class TypeVarInfo(NamedTuple):
...
+#
+# ChainMap does not exist in typing or typing_extensions in Python 2,
+# so we can disallow importing it from anywhere except collections
+_COLLECTIONS_NOT_TYPING_OR_TYPING_EXTENSIONS = {
+    "Counter": "Counter",
+    "Deque": "deque",
+    "DefaultDict": "defaultdict",
+    "ChainMap": "ChainMap",
+}
+_COLLECTIONS_NOT_TYPING_OR_TYPING_EXTENSIONS = {
```
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply pyi.py:57]: Okay, I'll see if I can refactor the dicts. They used to have shorter variable names, but @JelleZijlstra persuaded me to make them longer ;)

I _hate_ the fact that the dict values include `"` characters, but it's the best way I could see of doing it. The complicating factor is the `contextlib.ContextManager` error message -- if you try to import it from `typing_extensions`, the inserted snippet is `"contextlib.AbstractContextManager" (or "typing.ContextManager" in Python 2-compatible code)`. So the inserted snippets are not uniformly surrounded by `"` characters.

**Comment by AlexWaygood** [comment]
Typeshed PR to make the CI go green:

- https://github.com/python/typeshed/pull/6949

TODO:
- Try to refactor the dicts, per @Akuli's request
- Disallow importing `Type` from `typing_extensions` as well as `typing`

**Comment by AlexWaygood** [comment]
Ohhhh I hardcoded Y019 to look for `type[_T]` instead of `Type[_T]`, so changing all occurences of `Type` to `type` in Python 2 has caused the CI to pick up on a bunch of new Y019 errors.

**Comment by AlexWaygood** [comment]
flake8 is complaining that `PyiVisitor._check_import_or_attribute` is too complex for one function; however, I think it would make the code _more_ complex to split it up into several smaller functions, so I am going to `# noqa` it. Let me know if you disagree.

**Comment by AlexWaygood** [comment]
> flake8 is complaining that `PyiVisitor._check_import_or_attribute` is too complex for one function. However, I think it would make the code _more_ complex to split it up into several smaller functions, so I am going to `# noqa` it. Let me know if you disagree.

Actually, I'll leave it for now, since I might be refactoring that function anyway.

**Comment by AlexWaygood** [comment]
...And a third typeshed PR: https://github.com/python/typeshed/pull/6955

**Comment by AlexWaygood** [comment]
CI is finally green, and I think I've addressed the points @Akuli raised (just two dicts now, and managed to get rid of the icky quotes-within-quotes everywhere). Re-requesting review from you, @JelleZijlstra, if that's okay, as it's changed a fair bit since when you approved it :)

**Review by JelleZijlstra — APPROVED** [review]

**Review by Akuli — APPROVED** [review]
Skimmed. No red flags.

**Comment by AlexWaygood** [comment]
> Skimmed. No red flags.

Thanks!
